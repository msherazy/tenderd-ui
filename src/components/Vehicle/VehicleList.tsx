import React from 'react';
import type { Vehicle } from '../../types';
import { StatusBadge, VehicleTypeBadge } from '../Badge';
import { EmptyState, ErrorMessage, LoadingSpinner } from '../Card';
import { Index } from '../Button';
import { t } from '../../utils/locale';

interface VehicleListProps {
	vehicles: Vehicle[];
	loading: boolean;
	error: string | null;
	sortField: keyof Vehicle;
	sortDirection: 'asc' | 'desc';
	onVehicleSelect: (vehicle: Vehicle) => void;
	onSort: (field: keyof Vehicle) => void;
}

export const VehicleList: React.FC<VehicleListProps> = ({
	vehicles,
	loading,
	error,
	sortField,
	sortDirection,
	onVehicleSelect,
	onSort,
}) => {
	if (loading) return <LoadingSpinner />;
	if (error) return <ErrorMessage message={error} />;
	if (vehicles.length === 0) return <EmptyState />;

	const renderSortIndicator = (field: keyof Vehicle) => {
		if (sortField !== field) return null;
		return sortDirection === 'asc' ? '↑' : '↓';
	};

	return (
		<div className="overflow-x-auto">
			<table className="min-w-full bg-white divide-y divide-gray-200">
				<thead className="bg-gray-50">
					<tr>
						<th
							className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
							onClick={() => onSort('make')}
						>
							Make {renderSortIndicator('make')}
						</th>
						<th
							className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
							onClick={() => onSort('model')}
						>
							Model {renderSortIndicator('model')}
						</th>
						<th
							className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
							onClick={() => onSort('year')}
						>
							Year {renderSortIndicator('year')}
						</th>
						<th
							className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
							onClick={() => onSort('type')}
						>
							Type {renderSortIndicator('type')}
						</th>
						<th
							className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
							onClick={() => onSort('status')}
						>
							Status {renderSortIndicator('status')}
						</th>
						<th className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider">
							Actions
						</th>
					</tr>
				</thead>
				<tbody className="bg-white divide-y divide-gray-200">
					{vehicles?.map(vehicle => (
						<tr key={vehicle._id} className="hover:bg-gray-50 transition-colors">
							<td className="px-6 py-4 whitespace-nowrap text-sm font-medium text-gray-900">
								{vehicle.make}
							</td>
							<td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500">{vehicle.model}</td>
							<td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500">{vehicle.year}</td>
							<td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500">
								<VehicleTypeBadge type={vehicle.type} />
							</td>
							<td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500">
								<StatusBadge status={vehicle.status} />
							</td>
							<td className="px-6 py-4 whitespace-nowrap text-sm font-medium">
								<Index variant="primary" onClick={() => onVehicleSelect(vehicle)}>
									{t('VIEW_DETAILS')}
								</Index>
							</td>
						</tr>
					))}
				</tbody>
			</table>
		</div>
	);
};
