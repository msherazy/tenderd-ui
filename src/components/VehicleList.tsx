import React from 'react';
import type {Vehicle} from '../types';
import { StatusBadge, VehicleTypeBadge } from './Badges.tsx';
import { LoadingSpinner, ErrorMessage, EmptyState } from './Card.tsx';

interface VehicleListProps {
    vehicles: Vehicle[];
    loading: boolean;
    error: string | null;
    sortField: keyof Vehicle;
    sortDirection: 'asc' | 'desc';
    onVehicleSelect: (vehicle: Vehicle) => void;
    onSort: (field: keyof Vehicle) => void;
}

export const VehicleList: React.FC<VehicleListProps> = ({
    vehicles,
    loading,
    error,
    sortField,
    sortDirection,
    onVehicleSelect,
    onSort
}) => {
    if (loading) return <LoadingSpinner />;
    if (error) return <ErrorMessage message={error} />;
    if (vehicles.length === 0) return <EmptyState />;

    const renderSortIndicator = (field: keyof Vehicle) => {
        if (sortField !== field) return null;
        return sortDirection === 'asc' ? '↑' : '↓';
    };

    return (
        <div className="overflow-x-auto">
            <table className="min-w-full divide-y divide-gray-200 dark:divide-gray-700">
                <thead className="bg-gray-50 dark:bg-gray-800">
                    <tr>
                        <th
                            className="px-6 py-3 text-left text-xs font-medium text-gray-500 dark:text-gray-300 uppercase tracking-wider cursor-pointer"
                            onClick={() => onSort('make')}
                        >
                            Make {renderSortIndicator('make')}
                        </th>
                        <th
                            className="px-6 py-3 text-left text-xs font-medium text-gray-500 dark:text-gray-300 uppercase tracking-wider cursor-pointer"
                            onClick={() => onSort('model')}
                        >
                            Model {renderSortIndicator('model')}
                        </th>
                        <th
                            className="px-6 py-3 text-left text-xs font-medium text-gray-500 dark:text-gray-300 uppercase tracking-wider cursor-pointer"
                            onClick={() => onSort('year')}
                        >
                            Year {renderSortIndicator('year')}
                        </th>
                        <th
                            className="px-6 py-3 text-left text-xs font-medium text-gray-500 dark:text-gray-300 uppercase tracking-wider cursor-pointer"
                            onClick={() => onSort('type')}
                        >
                            Type {renderSortIndicator('type')}
                        </th>
                        <th
                            className="px-6 py-3 text-left text-xs font-medium text-gray-500 dark:text-gray-300 uppercase tracking-wider cursor-pointer"
                            onClick={() => onSort('status')}
                        >
                            Status {renderSortIndicator('status')}
                        </th>
                        <th className="px-6 py-3 text-left text-xs font-medium text-gray-500 dark:text-gray-300 uppercase tracking-wider">
                            Actions
                        </th>
                    </tr>
                </thead>
                <tbody className="bg-white dark:bg-gray-900 divide-y divide-gray-200 dark:divide-gray-700">
                    {vehicles.map(vehicle => (
                        <tr key={vehicle.id} className="hover:bg-gray-50 dark:hover:bg-gray-800">
                            <td className="px-6 py-4 whitespace-nowrap text-sm font-medium text-gray-900 dark:text-white">
                                {vehicle.make}
                            </td>
                            <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500 dark:text-gray-300">
                                {vehicle.model}
                            </td>
                            <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500 dark:text-gray-300">
                                {vehicle.year}
                            </td>
                            <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500 dark:text-gray-300">
                                <VehicleTypeBadge type={vehicle.type} />
                            </td>
                            <td className="px-6 py-4 whitespace-nowrap text-sm text-gray-500 dark:text-gray-300">
                                <StatusBadge status={vehicle.status} />
                            </td>
                            <td className="px-6 py-4 whitespace-nowrap text-sm font-medium">
                                <button
                                    onClick={() => onVehicleSelect(vehicle)}
                                    className="text-indigo-600 dark:text-indigo-400 hover:text-indigo-900 dark:hover:text-indigo-300"
                                >
                                    View Details
                                </button>
                            </td>
                        </tr>
                    ))}
                </tbody>
            </table>
        </div>
    );
};
