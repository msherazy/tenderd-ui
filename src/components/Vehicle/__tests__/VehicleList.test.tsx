import { render, screen } from '@testing-library/react';
import { VehicleList } from '../VehicleList';
import type { Vehicle } from '../../../types';


describe('VehicleList Component', () => {
  const mockVehicles: Vehicle[] = [
      {
          _id: 'vehicle-1',
          make: 'Toyota',
          model: 'Camry',
          year: 2022,
          type: 'Sedan',
          status: 'Active',
          mileage: 15000,
          licensePlate: 'ABC123',
          createdAt: '2022-01-01',
          updatedAt: '2022-06-01',
          lastServiceDate: '',
          color: '',
          fuelType: 'Electric',
          vin: '',
          purchaseDate: '',
          dailyUsage: [],
          weeklyUsage: []
      },
      {
          _id: 'vehicle-2',
          make: 'Honda',
          model: 'CR-V',
          year: 2021,
          type: 'SUV',
          status: 'Maintenance',
          mileage: 20000,
          licensePlate: 'XYZ789',
          createdAt: '2022-01-05',
          updatedAt: '2022-06-01',
          lastServiceDate: '',
          color: '',
          fuelType: 'Electric',
          vin: '',
          purchaseDate: '',
          dailyUsage: [],
          weeklyUsage: []
      },
  ];

  const defaultProps = {
    vehicles: mockVehicles,
    loading: false,
    error: null,
    sortField: 'make' as keyof Vehicle,
    sortDirection: 'asc' as const,
    onVehicleSelect: jest.fn(),
    onSort: jest.fn(),
  };

  test('renders empty state when vehicles array is empty', () => {
    render(<VehicleList {...defaultProps} vehicles={[]} />);
    expect(screen.getByTestId('empty-state')).toBeInTheDocument();
  });

  test('renders vehicle data in the table correctly', () => {
    render(<VehicleList {...defaultProps} />);

    // Check if the table headers are rendered - using getByRole to find table headers
    const headers = screen.getAllByRole('columnheader');
    expect(headers.length).toBeGreaterThanOrEqual(5); // At least 5 columns

    expect(headers.some(h => h.textContent?.includes('Make'))).toBeTruthy();
    expect(headers.some(h => h.textContent?.includes('Model'))).toBeTruthy();
    expect(headers.some(h => h.textContent?.includes('Year'))).toBeTruthy();
    expect(headers.some(h => h.textContent?.includes('Type'))).toBeTruthy();
    expect(headers.some(h => h.textContent?.includes('Status'))).toBeTruthy();

    expect(screen.getByText('Toyota')).toBeInTheDocument();
    expect(screen.getByText('Camry')).toBeInTheDocument();
    expect(screen.getByText('2022')).toBeInTheDocument();
    expect(screen.getByText('Honda')).toBeInTheDocument();
    expect(screen.getByText('CR-V')).toBeInTheDocument();
    expect(screen.getByText('2021')).toBeInTheDocument();
  });

  test('displays sort indicators correctly', () => {
    render(<VehicleList {...defaultProps} sortField="model" sortDirection="desc" />);
    const modelHeader = screen.getByText(/Model/);
    expect(modelHeader.textContent).toContain('â†“'); // Down arrow for descending
  });

  test('matches snapshot', () => {
    const { container } = render(<VehicleList {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  test('matches snapshot with error', () => {
    const { container } = render(<VehicleList {...defaultProps} error="Error loading vehicles" />);
    expect(container).toMatchSnapshot();
  });

  test('matches snapshot with empty vehicles array', () => {
    const { container } = render(<VehicleList {...defaultProps} vehicles={[]} />);
    expect(container).toMatchSnapshot();
  });
});
