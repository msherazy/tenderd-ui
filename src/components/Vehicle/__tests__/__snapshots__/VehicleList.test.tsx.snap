// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleList Component matches snapshot 1`] = `
<div>
  <div
    class="overflow-x-auto"
  >
    <table
      class="min-w-full bg-white divide-y divide-gray-200"
    >
      <thead
        class="bg-gray-50"
      >
        <tr>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
          >
            Make 
            â†‘
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
          >
            Model 
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
          >
            Year 
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
          >
            Type 
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider cursor-pointer hover:text-gray-700 transition-colors"
          >
            Status 
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody
        class="bg-white divide-y divide-gray-200"
      >
        <tr
          class="hover:bg-gray-50 transition-colors"
        >
          <td
            class="px-6 py-4 whitespace-nowrap text-sm font-medium text-gray-900"
          >
            Toyota
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
          >
            Camry
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
          >
            2022
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
          >
            <span
              class="inline-flex items-center px-2.5 py-0.5 rounded-full text-xs font-medium bg-blue-100 text-blue-800"
            >
              Sedan
            </span>
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
          >
            <span
              class="inline-flex items-center px-2.5 py-0.5 rounded-full text-xs font-medium bg-green-100 text-green-800"
            >
              Active
            </span>
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm font-medium"
          >
            <span
              class="text-primary hover:underline cursor-pointer"
            >
              View Details
            </span>
          </td>
        </tr>
        <tr
          class="hover:bg-gray-50 transition-colors"
        >
          <td
            class="px-6 py-4 whitespace-nowrap text-sm font-medium text-gray-900"
          >
            Honda
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
          >
            CR-V
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
          >
            2021
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
          >
            <span
              class="inline-flex items-center px-2.5 py-0.5 rounded-full text-xs font-medium bg-green-100 text-green-800"
            >
              SUV
            </span>
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
          >
            <span
              class="inline-flex items-center px-2.5 py-0.5 rounded-full text-xs font-medium bg-yellow-100 text-yellow-800"
            >
              Maintenance
            </span>
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm font-medium"
          >
            <span
              class="text-primary hover:underline cursor-pointer"
            >
              View Details
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`VehicleList Component matches snapshot with empty vehicles array 1`] = `
<div>
  <div
    class="text-center py-12"
    data-testid="empty-state"
  >
    <svg
      aria-hidden="true"
      class="mx-auto h-12 w-12 text-gray-400"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M9.172 16.172a4 4 0 015.656 0M9 10h.01M15 10h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        vector-effect="non-scaling-stroke"
      />
    </svg>
    <h3
      class="mt-2 text-sm font-medium text-gray-900"
    >
      No Vehicle found
    </h3>
    <p
      class="mt-1 text-sm text-gray-500"
    >
      Try adjusting your search or filter to find what you're looking for.
    </p>
    <div
      class="mt-6"
    >
      <button
        class="inline-flex items-center justify-center px-4 py-2 rounded font-semibold focus:outline-none transition disabled:opacity-60 w-auto hover:opacity-90 focus:ring-2 focus:ring-offset-2 focus:ring-primary inline-flex items-center"
        style="color: rgb(255, 255, 255);"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="-ml-1 mr-2 h-5 w-5"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 3a1 1 0 011 1v5h5a1 1 0 110 2h-5v5a1 1 0 11-2 0v-5H4a1 1 0 110-2h5V4a1 1 0 011-1z"
            fill-rule="evenodd"
          />
        </svg>
        Add Vehicle
      </button>
    </div>
  </div>
</div>
`;

exports[`VehicleList Component matches snapshot with error 1`] = `
<div>
  <div
    class="bg-red-50 border-l-4 border-red-500 p-4 my-6"
  >
    <div
      class="flex"
    >
      <div
        class="flex-shrink-0"
      >
        <svg
          class="h-5 w-5 text-red-500"
          fill="currentColor"
          viewBox="0 0 20 20"
        >
          <path
            clip-rule="evenodd"
            d="M10 18a8 8 0 100-16 8 8 0 000 16zM8.707 7.293a1 1 0 00-1.414 1.414L8.586 10l-1.293 1.293a1 1 0 101.414 1.414L10 11.414l1.293 1.293a1 1 0 001.414-1.414L11.414 10l1.293-1.293a1 1 0 00-1.414-1.414L10 8.586 8.707 7.293z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="ml-3"
      >
        <p
          class="text-sm text-red-700"
        >
          Error loading vehicles
        </p>
      </div>
    </div>
  </div>
</div>
`;
